lib/ssh/test/ssh_connection_SUITE.erl

```diff
@@ -1943,6 +1945,134 @@ max_channels_option(Config) when is_list(Config) ->
+            end,
+
+            Parent ! {self(), channelId, ChannelId0},
+            Result = receive
+                cmd ->
+                    ct:log("~p:~p Channel ~p executing", [?MODULE, ?LINE, ChannelId0]),
+                    success = ssh_connection:exec(ConnectionRef, ChannelId0, "testing", infinity),
+                    Expect = <<"echo testing\n">>,
+                    ExpSz = size(Expect),
+                    receive
+                        {ssh_cm, ConnectionRef, {data, ChannelId0, 0,
+                                                 <<Expect:ExpSz/binary, _/binary>>}} = R ->
+                            ct:log("~p:~p Got expected ~p",[?MODULE,?LINE, R]),
+                            ok;
+                        Other ->
+                            ct:log("~p:~p Got unexpected ~p~nExpect: ~p~n",
+                                   [?MODULE,?LINE, Other, {ssh_cm, ConnectionRef,
+                                                           {data, ChannelId0, 0, Expect}}]),
+                            {fail, "Unexpected data"}
+                    after 5000 ->
+                            {fail, "Exec Timeout"}
+                    end;
+                    stop -> {fail, "Stopped"}
+            end,
+            Parent ! {self(), Result}
+      end),
+    try
+        TestResult = receive
+            {ExecChannelPid, channelId, ExId} ->
+                ct:log("~p:~p Channel that should stay: ~p pid ~p", [?MODULE, ?LINE, ExId, ExecChannelPid]),
+                ConnPeer = receive {conn_peer, CM} -> CM end,
+                %% The sole purpose of this channel is to go down before the opening procedure is complete
+                DownChannelPid = spawn(
+                    fun() ->
+                        ct:log("~p:~p open channel (incomplete)",[?MODULE,?LINE]),
+                        Parent ! {self(), channelId, ok},
+                        %% This is to prevent the peer from answering our 'channel-open' in time
+                        sys:suspend(ConnPeer),
+                        {ok, _ChannelId} = ssh_connection:session_channel(ConnectionRef, infinity),
+                        ct:log("~p:~p open incomplete channel done - should not have happened",[?MODULE,?LINE]),
+                        Parent ! {self(), {fail, "Unexpected channel success"}}
+                    end),
+                MonRef = erlang:monitor(process, DownChannelPid),
+                receive
+                    {DownChannelPid, channelId, ok} ->
+                        ct:log("~p:~p Channel handler that won't continue: pid ~p", [?MODULE, ?LINE, DownChannelPid]),
+                        ensure_channels(ConnectionRef, 2),
+                        channel_down_sequence(DownChannelPid, ExecChannelPid, ExId, MonRef, ConnectionRef, ConnPeer)
+                end
+        end,
+        ensure_channels(ConnectionRef, 0)
+    after
+        ssh:close(ConnectionRef),
+        ssh:stop_daemon(Pid)
+    end.
+
+ensure_channels(ConnRef, Expected) ->
```

Comment:
I'm not sure I follow `ensure_channels/2` and `do_ensure_channels/3` functions.
- why having 2 functions for this? maybe not needed in this simple case ...
- what is the exit criteria if ChannelList does not reach expected number?

shouldn't we have some counter and do only certain number of iterations?

Reply:
If `ChannelList` does not reach the expected number the whole test is expected to be killed by the time trap. It is possible to limit the number of iterations, but I always find it hard to find a good value.
As for two functions - `ensure_channels` is meant to be a "short interface" and `do_ensure_channels` does the actual job.